simple/app.go

```diff
@@ -0,0 +1,334 @@
+package simple
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+
+	"github.com/grafana/grafana-app-sdk/app"
+	"github.com/grafana/grafana-app-sdk/k8s"
+	"github.com/grafana/grafana-app-sdk/operator"
+	"github.com/grafana/grafana-app-sdk/resource"
+)
+
+var _ app.Provider = &AppProvider{}
+
+type Converter k8s.Converter
+
+// AppProvider is a simple implementation of app.Provider which returns AppManifest when Manifest is called,
+// and calls NewAppFunc when NewApp is called.
+type AppProvider struct {
+	AppManifest app.Manifest
+	NewAppFunc  func(config app.Config) (app.App, error)
+}
+
+// Manifest returns the AppManifest in the AppProvider
+func (a *AppProvider) Manifest() app.Manifest {
+	return a.AppManifest
+}
+
+// NewApp calls NewAppFunc and returns the result
+func (a *AppProvider) NewApp(settings app.Config) (app.App, error) {
+	return a.NewAppFunc(settings)
+}
+
+// NewAppProvider is a convenience method for creating a new AppProvider
+func NewAppProvider(manifest app.Manifest, newAppFunc func(cfg app.Config) (app.App, error)) *AppProvider {
+	return &AppProvider{
+		AppManifest: manifest,
+		NewAppFunc:  newAppFunc,
+	}
+}
+
+var (
+	_ app.App = &App{}
+)
+
+// App is a simple, opinionated implementation of resource.App.
+// It must be created with NewApp to be valid.
+type App struct {
+	informerController *operator.InformerController
+	operator           *operator.Operator
+	clientGenerator    resource.ClientGenerator
+	kinds              map[string]AppManagedKind
+	internalKinds      map[string]resource.Kind
+	cfg                AppConfig
+	converters         map[string]Converter
+}
+
+// AppConfig is the configuration used by App
+type AppConfig struct {
+	Name           string
+	KubeConfig     rest.Config
+	InformerConfig AppInformerConfig
+	ManagedKinds   []AppManagedKind
+	Converters     map[schema.GroupKind]Converter
+}
+
+// AppInformerConfig contains configuration for the App's internal operator.InformerController
+type AppInformerConfig struct {
+	ErrorHandler       func(context.Context, error)
+	RetryPolicy        operator.RetryPolicy
+	RetryDequeuePolicy operator.RetryDequeuePolicy
+	FinalizerSupplier  operator.FinalizerSupplier
+}
+
+// AppManagedKind is a Kind and associated functionality used by an App.
+type AppManagedKind struct {
```

Comment:
Is there any distinction between a kind that the app manipulates vs one that it has a read-only relationship with? Is that distinction useful?

Reply:
I've introduced `simple.AppUnmanagedKind` and `simple.App.WatchKind(simple.AppUnmanagedKind)` for establishing watchers/reconcilers on kinds which the app does not manage (this can also still be done via the `AddRunnable` method, but it's more cumbersome).